File: packages/core/src/research_engine/domain/works.py
"""Authored works — the Phase 1 spine.

A work is rows from its first freeze: revisions own ordered blocks, blocks own
citation occurrences and links. Drafts validate before the database does; rows
are what came back.
"""

from __future__ import annotations

from datetime import datetime
from enum import StrEnum
from typing import Any
from uuid import UUID  # noqa: TC003 - pydantic needs it at runtime

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator


class WorkStatus(StrEnum):
    DRAFT = "draft"
    REVIEW = "review"
    PUBLISHED = "published"
    ARCHIVED = "archived"


class RevisionState(StrEnum):
    DRAFT = "draft"
    FROZEN = "frozen"
    PUBLISHED = "published"
    SUPERSEDED = "superseded"


class Placement(StrEnum):
    INLINE = "inline"
    BLOCK_END = "block_end"


class Edition(BaseModel):
    """One row of `bibliography.editions`: a Zotero key seen at ingest."""

    id: UUID
    zotero_key: str
    csl: dict[str, Any] = Field(default_factory=dict)
    created_at: datetime


class Work(BaseModel):
    id: UUID
    slug: str
    title: str
    work_type: str
    status: WorkStatus = WorkStatus.DRAFT
    language: str | None = None
    abstract: str | None = None
    current_revision_id: UUID | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)
    created_at: datetime
    updated_at: datetime
    archived_at: datetime | None = None


class WorkDraft(BaseModel):
    model_config = ConfigDict(extra="forbid")

    slug: str
    title: str
    work_type: str
    language: str | None = None
    abstract: str | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)

    @field_validator("slug", "title")
    @classmethod
    def _non_empty(cls, value: str) -> str:
        if not value.strip():
            raise ValueError("must be non-empty")
        return value


class WorkRevision(BaseModel):
    id: UUID
    work_id: UUID
    revision_number: int
    parent_revision_id: UUID | None = None
    state: RevisionState = RevisionState.DRAFT
    message: str | None = None
    content_hash: bytes | None = None
    created_by: str = "user"
    created_at: datetime
    frozen_at: datetime | None = None
    published_at: datetime | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)


class WorkRevisionDraft(BaseModel):
    model_config = ConfigDict(extra="forbid")

    work_id: UUID
    revision_number: int = 1
    parent_revision_id: UUID | None = None
    message: str | None = None
    created_by: str = "user"
    metadata: dict[str, Any] = Field(default_factory=dict)

    @field_validator("revision_number")
    @classmethod
    def _number_starts_at_one(cls, value: int) -> int:
        if value < 1:
            raise ValueError("revision_number starts at 1")
        return value


class WorkBlock(BaseModel):
    id: UUID
    revision_id: UUID
    block_key: UUID
    parent_id: UUID | None = None
    position: int
    block_type: str
    title: str | None = None
    body_markdown: str = ""
    attributes: dict[str, Any] = Field(default_factory=dict)
    created_at: datetime
    updated_at: datetime


class WorkBlockDraft(BaseModel):
    model_config = ConfigDict(extra="forbid")

    revision_id: UUID
    block_key: UUID
    parent_id: UUID | None = None
    position: int = 0
    block_type: str
    title: str | None = None
    body_markdown: str = ""
    attributes: dict[str, Any] = Field(default_factory=dict)

    @field_validator("position")
    @classmethod
    def _position_is_an_order(cls, value: int) -> int:
        if value < 0:
            raise ValueError("position is a 0-based order, never negative")
        return value


class BlockSourceLink(BaseModel):
    block_id: UUID
    source_span_id: UUID
    relation: str
    confidence: float | None = None
    note: str | None = None
    created_at: datetime


class BlockSourceLinkDraft(BaseModel):
    model_config = ConfigDict(extra="forbid")

    block_id: UUID
    source_span_id: UUID
    relation: str
    confidence: float | None = None
    note: str | None = None

    @field_validator("confidence")
    @classmethod
    def _confidence_is_a_fraction(cls, value: float | None) -> float | None:
        if value is not None and not 0 <= value <= 1:
            raise ValueError("confidence must be between 0 and 1")
        return value


class BlockEntityLink(BaseModel):
    block_id: UUID
    entity_id: UUID
    relation: str
    surface_form: str | None = None
    created_at: datetime


class BlockLinks(BaseModel):
    """Both link kinds of one block, for get and trace."""

    sources: list[BlockSourceLink] = Field(default_factory=list)
    entities: list[BlockEntityLink] = Field(default_factory=list)


class BlockEntityLinkDraft(BaseModel):
    model_config = ConfigDict(extra="forbid")

    block_id: UUID
    entity_id: UUID
    relation: str
    surface_form: str | None = None


class Waiver(BaseModel):
    """A row that lets a gated finding pass: who, what, why."""

    id: UUID
    revision_id: UUID
    rule_id: str
    subject: str | None = None
    actor: str
    reason: str
    created_at: datetime


class WaiverDraft(BaseModel):
    model_config = ConfigDict(extra="forbid")

    revision_id: UUID
    rule_id: str
    subject: str | None = None
    actor: str = "user"
    reason: str

    @field_validator("rule_id", "reason")
    @classmethod
    def _non_empty(cls, value: str) -> str:
        if not value.strip():
            raise ValueError("must be non-empty")
        return value

    @model_validator(mode="after")
    def _actor_is_named(self) -> WaiverDraft:
        if not self.actor.strip():
            raise ValueError("a waiver records who, not just why")
        return self

File: crates/marginalia-types/src/works.rs
//! Authored-works spine mirroring `domain/works.py`.

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use serde_json::{Map, Value};
use uuid::Uuid;

use crate::errors::{Error, Result};

#[derive(Debug, Clone, Copy, Default, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum WorkStatus {
    #[default]
    Draft,
    Published,
    Archived,
}

#[derive(Debug, Clone, Copy, Default, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum RevisionState {
    #[default]
    Draft,
    Frozen,
    Published,
    Superseded,
}

#[derive(Debug, Clone, Copy, Default, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum Placement {
    #[default]
    Inline,
    BlockEnd,
}

/// One row of `bibliography.editions`: an edition key seen at ingest.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct Edition {
    pub id: Uuid,
    pub edition_key: String,
    #[serde(default)]
    pub csl: Map<String, Value>,
    pub created_at: DateTime<Utc>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct Work {
    pub id: Uuid,
    pub slug: String,
    pub title: String,
    pub work_type: String,
    #[serde(default)]
    pub status: WorkStatus,
    #[serde(default)]
    pub language: Option<String>,
    #[serde(default, rename = "abstract")]
    pub abstract_text: Option<String>,
    #[serde(default)]
    pub current_revision_id: Option<Uuid>,
    #[serde(default)]
    pub metadata: Map<String, Value>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
    #[serde(default)]
    pub archived_at: Option<DateTime<Utc>>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct WorkDraft {
    pub slug: String,
    pub title: String,
    pub work_type: String,
    #[serde(default)]
    pub language: Option<String>,
    #[serde(default, rename = "abstract")]
    pub abstract_text: Option<String>,
    #[serde(default)]
    pub metadata: Map<String, Value>,
}

impl WorkDraft {
    pub fn validate(&self) -> Result<()> {
        if self.slug.trim().is_empty() || self.title.trim().is_empty() {
            return Err(Error::Validation("must be non-empty".to_owned()));
        }
        Ok(())
    }
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct WorkRevision {
    pub id: Uuid,
    pub work_id: Uuid,
    pub revision_number: i64,
    #[serde(default)]
    pub parent_revision_id: Option<Uuid>,
    #[serde(default)]
    pub state: RevisionState,
    #[serde(default)]
    pub message: Option<String>,
    #[serde(default)]
    #[serde(with = "crate::wire::bytes_string_opt")]
    pub content_hash: Option<Vec<u8>>,
    #[serde(default = "default_created_by")]
    pub created_by: String,
    pub created_at: DateTime<Utc>,
    #[serde(default)]
    pub frozen_at: Option<DateTime<Utc>>,
    #[serde(default)]
    pub published_at: Option<DateTime<Utc>>,
    #[serde(default)]
    pub metadata: Map<String, Value>,
}

fn default_created_by() -> String {
    "user".to_owned()
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct WorkRevisionDraft {
    pub work_id: Uuid,
    #[serde(default = "default_revision_number")]
    pub revision_number: i64,
    #[serde(default)]
    pub parent_revision_id: Option<Uuid>,
    #[serde(default)]
    pub message: Option<String>,
    #[serde(default = "default_created_by")]
    pub created_by: String,
    #[serde(default)]
    pub metadata: Map<String, Value>,
}

fn default_revision_number() -> i64 {
    1
}

impl WorkRevisionDraft {
    pub fn validate(&self) -> Result<()> {
        if self.revision_number < 1 {
            return Err(Error::Validation("revision_number starts at 1".to_owned()));
        }
        Ok(())
    }
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct WorkBlock {
    pub id: Uuid,
    pub revision_id: Uuid,
    pub block_key: Uuid,
    #[serde(default)]
    pub parent_id: Option<Uuid>,
    pub position: i64,
    pub block_type: String,
    #[serde(default)]
    pub title: Option<String>,
    #[serde(default)]
    pub body_markdown: String,
    #[serde(default)]
    pub attributes: Map<String, Value>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct WorkBlockDraft {
    pub revision_id: Uuid,
    pub block_key: Uuid,
    #[serde(default)]
    pub parent_id: Option<Uuid>,
    #[serde(default)]
    pub position: i64,
    pub block_type: String,
    #[serde(default)]
    pub title: Option<String>,
    #[serde(default)]
    pub body_markdown: String,
    #[serde(default)]
    pub attributes: Map<String, Value>,
}

impl WorkBlockDraft {
    pub fn validate(&self) -> Result<()> {
        if self.position < 0 {
            return Err(Error::Validation(
                "position is a 0-based order, never negative".to_owned(),
            ));
        }
        Ok(())
    }
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct BlockSourceLink {
    pub block_id: Uuid,
    pub source_span_id: Uuid,
    pub relation: String,
    #[serde(default)]
    pub confidence: Option<f64>,
    #[serde(default)]
    pub note: Option<String>,
    pub created_at: DateTime<Utc>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct BlockSourceLinkDraft {
    pub block_id: Uuid,
    pub source_span_id: Uuid,
    pub relation: String,
    #[serde(default)]
    pub confidence: Option<f64>,
    #[serde(default)]
    pub note: Option<String>,
}

impl BlockSourceLinkDraft {
    pub fn validate(&self) -> Result<()> {
        if let Some(c) = self.confidence {
            check_fraction(c)?;
        }
        Ok(())
    }
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct BlockEntityLink {
    pub block_id: Uuid,
    pub entity_id: Uuid,
    pub relation: String,
    #[serde(default)]
    pub surface_form: Option<String>,
    pub created_at: DateTime<Utc>,
}

/// Both link kinds of one block, for get and trace.
#[derive(Debug, Clone, Default, PartialEq, Serialize, Deserialize)]
pub struct BlockLinks {
    #[serde(default)]
    pub sources: Vec<BlockSourceLink>,
    #[serde(default)]
    pub entities: Vec<BlockEntityLink>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct BlockEntityLinkDraft {
    pub block_id: Uuid,
    pub entity_id: Uuid,
    pub relation: String,
    #[serde(default)]
    pub surface_form: Option<String>,
}

/// A row that lets a gated finding pass: who, what, why.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct Waiver {
    pub id: Uuid,
    pub revision_id: Uuid,
    pub rule_id: String,
    #[serde(default)]
    pub subject: Option<String>,
    pub actor: String,
    pub reason: String,
    pub created_at: DateTime<Utc>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct WaiverDraft {
    pub revision_id: Uuid,
    pub rule_id: String,
    #[serde(default)]
    pub subject: Option<String>,
    #[serde(default = "default_actor")]
    pub actor: String,
    pub reason: String,
}

fn default_actor() -> String {
    "user".to_owned()
}

impl WaiverDraft {
    pub fn validate(&self) -> Result<()> {
        if self.rule_id.trim().is_empty() || self.reason.trim().is_empty() {
            return Err(Error::Validation("must be non-empty".to_owned()));
        }
        if self.actor.trim().is_empty() {
            return Err(Error::Validation(
                "a waiver records who, not just why".to_owned(),
            ));
        }
        Ok(())
    }
}

fn check_fraction(value: f64) -> Result<()> {
    if !(0.0..=1.0).contains(&value) {
        return Err(Error::Validation(
            "confidence must be between 0 and 1".to_owned(),
        ));
    }
    Ok(())
}
